RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,547 @@
+    T* row_buf_workspace = workspace_buf + row * workspace_ld;
+
+    // Apply the original FHT on this row
+    for (int64_t i = 0; i < log_n; ++i) {
+        int64_t s1 = 1 << i;
+        int64_t s2 = s1 << 1;
+        for (int64_t j = 0; j < n; j += s2) {
+            for (int64_t k = 0; k < s1; ++k) {
+                bool b1 = j + k < num_cols;
+                bool b2 = j + k + s1 < num_cols;
+                T u = b1 ? row_buf[j + k] : row_buf_workspace[j + k - num_cols];
+                T v = b2 ? row_buf[j + k + s1] : row_buf_workspace[j + k + s1 - num_cols];
+                if(b1 && b2) {
+                    row_buf[j + k] = u + v;
+                    row_buf[j + k + s1] = u - v;
+                }
+                else if(!b2 && b1) {
+                    row_buf[j + k] = u + v;
+                    row_buf_workspace[j + k + s1 - num_cols] = u - v;
+                }
+                else if(!b2 && !b1) {
+                    row_buf_workspace[j + k - num_cols] = u + v;
+                    row_buf_workspace[j + k + s1 - num_cols] = u - v;
+                }
+            }
+        }
+    }
+}
+}
+
+template <typename T>
+void fht_right_col_major(T *buf, T* workspace_buf, int64_t log_n, int64_t num_rows, int64_t num_cols) {
+int64_t n = 1 << log_n;
+
+// Apply FHT to each row independently
+for (int64_t row= 0; row < num_rows; ++row) {
+    // Apply the original FHT on this row
+    for (int64_t i = 0; i < log_n; ++i) {
+        int64_t s1 = 1 << i;
+        int64_t s2 = s1 << 1;
+        for (int64_t j = 0; j < n; j += s2) {
+            for (int64_t k = 0; k < s1; ++k) {
+                bool b1 = (j + k) * num_rows + row < num_cols * num_rows;
+                bool b2 = (j + k + s1) * num_rows + row < num_cols * num_rows;
+                T u = b1 ? buf[(j + k) * num_rows + row] : workspace_buf[(j + k) * num_rows + row - num_rows * num_cols];
+                T v = b2 ? buf[(j + k + s1) * num_rows + row] : workspace_buf[(j + k + s1) * num_rows + row - num_rows * num_cols];
+                if(b1 && b2) {
+                    buf[(j + k) * num_rows + row] = u + v;
+                    buf[(j + k + s1) * num_rows + row] = u - v;
+                }
+                else if(!b2 && b1) {
+                    buf[(j + k) * num_rows + row] = u + v;
+                    workspace_buf[(j + k + s1) * num_rows + row - num_rows * num_cols] = u - v;
+                }
+                else if(!b2 && !b1) {
+                    workspace_buf[(j + k) * num_rows + row - num_rows * num_cols] = u + v;
+                    workspace_buf[(j + k + s1) * num_rows + row - num_rows * num_cols] = u - v;
+                }
+            }
+        }
+    }
+}
+}
+
+template <typename T>
+void fht_dispatch(
+    bool left, // Pre-multiplying?
+    blas::Layout layout,
+    int64_t num_rows,
+    int64_t num_cols,
+    int64_t log_n,
+    T* A,
+    int64_t workspace_ld=0, // leading dimension of the workspace buffer
+    T* workspace_buf=nullptr
```

Comment:
If ``workspace_buf`` is nullptr then this can segfault. Either make it a required argument or have special logic to handle when it's nullptr.